File: Greeter.java
public final class Greeter {
    public Greeter(String greeting) {
        initialize(this, greeting);
    }

    private Greeter(long peer) {
        this.peer = peer;
    }

    public native String greeting(String name);

    private native void initialize(Greeter self, String greeting);
    protected native void finalize() throws Throwable;
    private long peer;
}

File: greeter.cpp
#include "greeter.hpp"
#include "greeter_impl.hpp"

#include <utility>

Greeter::Greeter(std::string greeting)
: impl(std::make_unique<GreeterImpl>(std::move(greeting)))
{}

Greeter::~Greeter() = default;

std::string Greeter::greeting(const std::string& name) {
    return impl->greeting(name);
}
